linode/terraform-provider-linode
# new: Overhaul node pool update logic; make count optional with an autoscaler

## 📝 Description

This change overhauls the `linode_lke_cluster` node pool update logic to properly make changes reflected in the Terraform plan. Additionally, this PR adjusts the pool-reordering logic in the read function to only select on ID first followed by type and count. This should _hopefully_ prevent broken diffs going forward.

Previously, the node pool update logic worked by calculating diffs between the planned pools and the existing pools in the update routine. This functionality was both inconsistent and unpredictable, which has led to a lot of broken diffs and unexpected changes for users.

Unfortunately there is a caveat to the new approach: pool order is now significant because we're relying on ID values from state. This can be a bit annoying for users looking to drop a single pool (e.g. the first defined pool) because the provider will drop the last pool and resize all subsequent pools. I looked into implementing the ID-based reordering without overhauling the update logic but the unpredictable nature of the old update logic kept breaking everything.

This PR also allows users to exclude the `count` attribute from a `linode_lke_cluster.pool` if an autoscaler is defined. This change is necessary because users may not want to override their node count on every apply, and relying on the `lifecycle.ignore_changes` rule can have confusing effects on diffing.
 
Please try your hardest to break the diffing when updating node pools, we need this implementation to be as reliable as possible 🙂 

## ✔️ How to Test

The following test steps assume you have pulled down this change locally.

### E2E Testing

```bash
make PKG_NAME=linode/lke int-test
```

### Unit Testing

```bash
make unit-test
```

### Manual Testing

The following test steps assume you are using a provider sandbox environment (e.g. dx-devenv).

#### Testing Changes to Diffing and Updates

1. Apply an arbitrary LKE cluster:

```terraform

resource "linode_lke_cluster" "test-cluster" {
  label       = "my-cluster"
  k8s_version = "1.28"
  region      = "us-ord"

  pool {
    type = "g6-standard-1"
  }

  pool {
    type = "g6-standard-1"
    count = 2

    autoscaler {
      max = 3
      min = 2
    }
  }
}
```

2. Make any number of arbitrary changes to the node pool configuration.
3. Apply the configuration and ensure the planned changes are sensible.
4. Re-apply the configuration and ensure no changes are planned.

#### Testing Changes to the Count Field

1. Apply the following configuration:

```terraform
resource "linode_lke_cluster" "test-cluster" {
  label       = "my-cluster"
  k8s_version = "1.28"
  region      = "us-ord"

  pool {
    type = "g6-standard-1"
  }

  pool {
    type = "g6-standard-1"
    count = 2

    autoscaler {
      max = 3
      min = 2
    }
  }
}
```

5. Ensure the following error is raised:

```
pool.0: `count` must be defined when no autoscaler is defined
```

6. Add an autoscaler to the first node pool and apply:

```terraform
resource "linode_lke_cluster" "test-cluster" {
  label       = "my-cluster"
  k8s_version = "1.28"
  region      = "us-ord"

  pool {
    type = "g6-standard-1"

    autoscaler {
      max = 3
      min = 1
    }
  }

  pool {
    type = "g6-standard-1"
    count = 2

    autoscaler {
      max = 3
      min = 2
    }
  }
}
```

7. Observe Terraform successfully provisions the cluster as expected.
8. Navigate to the cluster in Cloud Manager. Ensure the cluster has two node pools:

- Linode 2GB; 1 node; Autoscaler 1-3
- Linode 2GB; 2 nodes; Autoscaler 2-3

9. In Cloud Manager, manually apply the following resizes:

- `Linode 2GB; 1 node; Autoscaler 1-3` => 2 nodes
- `Linode 2GB; 2 nodes; Autoscaler 2-3` => 3 nodes

10. Reapply the configuration from step 3.
11. Observe that no changes are planned for the pool without an explicit count and that the pool with a count is planned to be downsized.
12. Observe that all changes are reflected in Cloud Manager.
